### PR TITLE
Add plan and function extension support

### PR DESCRIPTION
**Pull request by nseekhao**
## Details

- Add support for producing and consuming `Substrait` `Plan`s (as opposed to `Rel`)
- Add support for extension functions

### CODE REVIEW

File: src/consumer.rs [hunk 1]
```diff
@@ -60,6 +61,27 @@ pub fn reference_to_op(reference: u32) -> Result<Operator> {
     }
 }
 
+/// Convert Substrait Plan to DataFusion DataFrame
+pub async fn from_substrait_plan(ctx: &mut SessionContext, plan: &Plan) -> Result<Arc<DataFrame>> {
+    match plan.relations.len() {
+        1 => {
+            match plan.relations[0].rel_type.as_ref().unwrap() {
```
Comment (andygrove) [inline src/consumer.rs:68]: It would be nice to remove the `unwrap` here

File: src/producer.rs [hunk 2]
```diff
@@ -18,11 +18,37 @@ use substrait::protobuf::{
...
+    Expression, FetchRel, FilterRel, FunctionArgument, JoinRel, NamedStruct, ProjectRel, ReadRel, SortField, SortRel,
+    PlanRel,
+    Plan, Rel, extensions::{self, simple_extension_declaration::{MappingType, ExtensionFunction}},
 };
 
+/// Convert DataFusion LogicalPlan to Substrait Plan
+pub fn to_substrait_plan(plan: &LogicalPlan) -> Result<Box<Plan>> {
+    // Parse relation nodes
+    let mut extensions: Vec<extensions::SimpleExtensionDeclaration> = vec![];
+    let proto = to_substrait_rel(plan, &mut extensions).unwrap();
```
Comment (andygrove) [inline src/producer.rs:30]: Same comment re `unwrap`

**Review by andygrove — APPROVED** [review]
LGTM. Would be nice to remove the `unwrap`s.

**Comment by nseekhao** [comment]
> LGTM. Would be nice to remove the `unwrap`s.

👍 Done.

**Review by andygrove — APPROVED** [review]

**Pull request merged**